Streaming

jcushman opened the pull request.

jcushman committed:
Stream responses

jcushman committed:
Switch default to gpt-4 turbo

rebeccacremona commented on aibot.py:
```diff
@@ -330,9 +341,23 @@ def handle_conversation(say, payload, is_dm=False):
-        response = get_text(prompt_messages)
-        response = response.rsplit('[name_separator]', 1)[-1]
-        say(response, response_type="in_channel")
+        # this should stream the results to the user by posting and editing a message ...
```
Wicked cool!

Do you think Abby could revert to her prior behavior, if a person asked politely using a particular syntax?

matteocargnelutti commented:
@jcushman Cool! 

Random suggestion: It might be worth upgrading the `openai` client (version used here:  `0.27.2`, latest: `1.16.1`). 

There is a [migration tool](https://github.com/openai/openai-python/discussions/742) available, but _I think_ the only thing you'd actually have to change is this line: 

https://github.com/jcushman/aibot/blob/359574ca81c7bbde454f27bc7fed793e4a190ba8/aibot.py#L68 

```python3
# Before
response = openai.ChatCompletion.create(...)
# After 
response = openai.chat.completion.create(...)
```

jcushman committed:
Update dependencies

jcushman committed:
Customizable streaming speed

jcushman commented:
Updated all dependencies by adding pip-tools; I'd be fine with another approach if Ben cares but that was the easiest for me.

Added a Slack setting for faster or slower speeding, which can be set to zero to disable streaming.

<img width="494" alt="image" src="https://github.com/harvard-lil/aibot/assets/376272/544d87de-7673-4309-8c47-2efa61a61c62">

This is a little wonky because it's based on the setting of the user who sent the most recent message, not your settings as a viewer; but it'll work when talking directly to the App rather than in a channel at least, which seems to be the most useful mode.

bensteinberg reviewed (approved):
Nice!
